File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque


# 시작, 레버, 탈출 위치 찾기
def find_position(maps):
    positions = {'S': [], 'E': [], 'L': []}

    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if maps[i][j] in positions:
                positions[maps[i][j]].append((j, i))

    return positions


def bfs(start, end, maps):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    l = len(maps)  # 세로
    w = len(maps[0])  # 가로
    visited = [[False for _ in range(w)] for _ in range(l)]
    d = deque()

    d.append((start[0][0], start[0][1], 0))

    while d:
        x, y, cnt = d.popleft()
        
        if x == end[0][0] and y == end[0][1]:
            return cnt

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < w and 0 <= ny < l and maps[ny][nx] != 'X':
                if not visited[ny][nx]:  # 방문하지 않았으면
                    d.append((nx, ny, cnt + 1))
                    visited[ny][nx] = True

    return -1  # 갈 수 없으면


def solution(maps):
    answer = -1
    positions = find_position(maps)

    path1 = bfs(positions['S'], positions['L'], maps)  # start -> lever
    path2 = bfs(positions['L'], positions['E'], maps)  # lever-> exit
    if path1 != -1 and path2 != -1:
        answer = path1 + path2

    return answer

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    public int solution(String[] maps) {
        // 시작지점, 레버, exit 찾기
        Pos start = new Pos();
        Pos lev = new Pos();
        Pos exit = new Pos();
        
        for(int i = 0; i < maps.length; i++){
            for(int j = 0; j < maps[i].length(); j++){
                if(maps[i].charAt(j) == 'S'){
                    start = new Pos(i, j, 0);
                } else if(maps[i].charAt(j) == 'L') {
                    lev = new Pos(i, j, 0);
                } else if(maps[i].charAt(j) == 'E') {
                    exit = new Pos(i, j, 0);
                }
            }
        }
        
        
        // 시작지점 -> 레버
        int d1 = bfs(maps, start, lev);
        if(d1 == -1) return -1;
            
        // 레버 -> exit 
        int d2 = bfs(maps, lev, exit);
        if(d2 == -1) return -1;

        return d1 + d2;
    }
    
    public int bfs(String[] maps, Pos start, Pos end){
        Queue<Pos> queue = new LinkedList<>();
        int n = maps.length;
        int m = maps[0].length();
        boolean[][] visited = new boolean[n][m];
        int[] dx = {0, 0, -1, 1};
        int[] dy = {1, -1, 0, 0};
        
        queue.add(start);
        visited[start.x][start.y] = true;
        
        while(!queue.isEmpty()){
            Pos now = queue.poll();
            int x = now.x;
            int y = now.y;
            // System.out.println(now.x + " " + now.y);
            
            if(now.x == end.x && now.y == end.y){ // 도착하면
                return now.dist;
            }
            
            for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(nx >= 0 && nx < n && ny >=0 && ny < m){
                    if(!visited[nx][ny] && maps[nx].charAt(ny) != 'X'){
                        visited[nx][ny] = true;
                        queue.add(new Pos(nx, ny, now.dist + 1));
                    }
                }
            }
            
        }
        
        return -1;
        
    }
    
    
    
    public class Pos {
        int x;
        int y;
        int dist;
        
        public Pos(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
        
        public Pos() {
        }
    }
}